init

Helgapolis opened the pull request:
Создан плеер для проигрывания музыки YouTube 
Для запросов использовался открытый сервис https://openwhyd.github.io/openwhyd/
Код написан полностью на котлине
Для загрузки контента использовались корутины, можно посмотреть в пакетах category и trackList 
Для запросов к сервису использовался Retrofit, можно посмотреть в пакете trackList
Был добавлен пока небольшой домейн слой(пока довольно примитивный с получением данных из репозитория)
В качестве основного презентационного слоя использовался MVVM, но в адаптере пока используется MVP(думаю некоторое смешение стилей допустимо?, возможно в адаптере после сделаю изменения)

Olga1987 committed:
init

Olga1987 committed:
adding of new category is added

darja commented on app/build.gradle:
```diff
@@ -0,0 +1,64 @@
+    def retrofit2_version = "2.5.0"
+    def okhttp3_version = "3.12.0"
+    def kotlinCoroutineVersion = "1.0.1"
+    def nav_version = "1.0.0"
```
У такого способа указания версий есть один минус: Android Studio их не парсит и не сообщает об обновлениях. Если версии указаны явно, вроде `implementation 'androidx.core:core-ktx:1.2.0`, то при выходе обновлений появится warning

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/data/TrackCategoryRepository.kt:
```diff
@@ -0,0 +1,5 @@
+package ru.akulinina.musicplayer.category.data
+
+interface TrackCategoryRepository {
```
Интерфейсы репозиториев относятся к слою domain, а не data. Иначе получается, что domain имеет ссылку на data, что неправильно.

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/CategoryFragment.kt:
```diff
@@ -0,0 +1,66 @@
+
+        val layoutManager = LinearLayoutManager(view.context)
+        fragmentTrackCategoryContentList.layoutManager = layoutManager
+        val dividerItemDecoration = DividerItemDecoration(view.context, layoutManager.orientation)
```
Немногие знают, но divider-ы в андроиде не то чтобы сильно приветствуются. Их главное назначение – группировка, а не разделение. В [гайдлайнах](https://material.io/components/dividers/#usage) сказано Use dividers sparingly, to create groupings rather than separate items. В данном случае я дивайдеры бы не использовала.
И стрелочки справа – паттерн из iOS, который тоже многие копируют в андроиде, хотя в гайдлайнах нигде об этом не написано. К слову, гугл, когда проверяет приложения для фичеринга, прямо заставляет эти стрелочки убирать.

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/TrackCategoryAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+
+class TrackCategoryAdapter(private val trackCategoryRouter: TrackCategoryRouter?): RecyclerView.Adapter<TrackCategoryViewHolder>() {
+
+    @Suppress("SENSELESS_COMPARISON")
```
Советую писать к `@Suppress` комментарии

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/TrackCategoryPresenter.kt:
```diff
@@ -0,0 +1,28 @@
+package ru.akulinina.musicplayer.category.presentation
+
+class TrackCategoryPresenter(private var model: String) {
```
Все классы обвязки для RecyclerViewAdapter можно сделать internal, поскольку они все живут в одном пакете и снаружи не нужны.

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/TrackCategoryPresenter.kt:
```diff
@@ -0,0 +1,28 @@
+        this.view?.setName(model)
+    }
+
+    fun detachView() {
```
Неиспользуемые функции лучше удалять

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/TrackCategoryViewModelProvider.kt:
```diff
@@ -0,0 +1,13 @@
+import androidx.lifecycle.ViewModelProvider
+import ru.akulinina.musicplayer.category.data.TrackCategoryRepository
+
+class TrackCategoryViewModelProvider(private val repository: TrackCategoryRepository) : ViewModelProvider.Factory {
```
Интерфейс называется `Factory`, так что лучше называть реализации согласованно, `TrackCategoryViewModelFactory`

darja commented on app/src/main/java/ru/akulinina/musicplayer/track/presentation/TrackVideoActivity.kt:
```diff
@@ -0,0 +1,45 @@
+class TrackVideoActivity : YouTubeBaseActivity(), YouTubePlayer.OnInitializedListener {
+
+    companion object{
+        const val KEY_INTENT_TRACK = "key_intent_track"
```
Ключи для Extras принято называть с префикса `EXTRA_`

darja commented on app/src/main/java/ru/akulinina/musicplayer/track/presentation/TrackVideoActivity.kt:
```diff
@@ -0,0 +1,45 @@
+        setContentView(R.layout.activity_track)
+
+        val track = intent?.getSerializableExtra(KEY_INTENT_TRACK) as? Track
+        val trackIdItems = track?.eId?.split("/")
+        trackId = trackIdItems?.last() ?: ""
```
Кажется, можно объединить в одно выражение

darja commented on app/src/main/java/ru/akulinina/musicplayer/track/presentation/TrackVideoActivity.kt:
```diff
@@ -0,0 +1,45 @@
+    override fun onInitializationSuccess(
+        provider: YouTubePlayer.Provider?,
+        youTubePlayer: YouTubePlayer?,
+        b: Boolean
```
Имя `b` ничего не говорит о назначении этой переменной. Возможно, там на самом деле что-то такое, что стоило бы проверить

darja commented on app/src/main/java/ru/akulinina/musicplayer/track/presentation/TrackVideoActivity.kt:
```diff
@@ -0,0 +1,45 @@
+        provider: YouTubePlayer.Provider?,
+        youTubePlayer: YouTubeInitializationResult?
+    ) {
+        /* do nothing */
```
Тут бы не помешали логи

darja commented on app/src/main/java/ru/akulinina/musicplayer/tracklist/data/RetrofitAPI.kt:
```diff
@@ -0,0 +1,34 @@
+
+interface TrackApiService {
+    @GET("search?format=json")
+    fun getTracksAsync(@Query("q") query: String): Deferred<Response<SearchResponse>>
```
Я бы использовала suspend вместо Deferred

darja commented on app/src/main/java/ru/akulinina/musicplayer/tracklist/domain/GetTrackListUseCase.kt:
```diff
@@ -0,0 +1,7 @@
+
+import ru.akulinina.musicplayer.tracklist.dto.Track
+
+interface GetTrackListUseCase {
```
На мой взгляд, в подавляющем большинстве случаев юскейсам не нужны интерфейсы. Бывают случаи, когда нужны два юскейса с одной сигнатурой, но это прямо редкость.

darja commented on app/src/main/java/ru/akulinina/musicplayer/tracklist/domain/GetTrackListMainUseCase.kt:
```diff
@@ -0,0 +1,11 @@
+class GetTrackListMainUseCase(private val repository: TracksRepository) : GetTrackListUseCase {
+
+    override suspend fun getTrackList(query: String): List<Track> {
+        return repository.getTracks(query) ?: listOf()
```
Заводить ли юскейс, если в нём будет только вызов метода репозитория – философский вопрос. С одной стороны, ходить сразу в репозиторий как-то нехорошо. С другой  стороны, городить кучу однострочных юскейсов тоже не имеет особенно смысла. Я для себя этот вопрос решаю так: если есть что тестировать – тогда юскейс. Если нечего – то вызов прямо из репозитория.

darja commented on app/src/main/java/ru/akulinina/musicplayer/tracklist/presentation/TrackListFragment.kt:
```diff
@@ -0,0 +1,71 @@
+class TrackListFragment : Fragment(), TrackItemRouter {
+
+    companion object {
+        const val KEY_BUNDLE_QUERY = "key_bundle_query"
```
Ключи для аргументов я бы называла с `ARG_`. `KEY_` – слишком общий префикс

darja commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,25 @@
+    </style>
+
+    <!-- TextView -->
+    <style name="text_view">
```
Стили в Андроиде названы в UpperCamelCase. Рекомендую следовать этому правилу

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/TrackCategoryAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+            field = value
+            presenters.clear()
+
+            for (track in tracks) {
+                if (track == null) {
+                    continue
+                }
+
+                val presenter = TrackCategoryPresenter(track)
+                presenters.add(presenter)
+            }
```
В котлине этот цикл можно заменить одной строчкой:
```
presenters.addAll(tracks.map { TrackCategoryPresenter(it) })
```
Null-check, кажется, вообще не нужен: в `tracks` содержатся String, а не String?

darja commented on app/src/main/java/ru/akulinina/musicplayer/category/presentation/TrackCategoryAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+    @Suppress("SENSELESS_COMPARISON")
+    var tracks: List<String> = listOf()
+        set(value) {
+            field = value
+            presenters.clear()
```
Котлин позволяет выносить логику в сеттеры, но я бы не стала так делать. От строчки вида `adapter.tracks = newTracks` ожидаешь только присваивания, а не обновления списка. Я бы объявила функцию типа `updateItems` и использовала её вместо сеттера.

Helgapolis committed:
gradle,sdk versions are updated, deprecated kotlin android extensions are replaced to ViewBinding

Helgapolis committed:
The dependency is removed(Starting from Kotlin 1.4 dependency on the standard library added by default)

Helgapolis committed:
The dependency is removed(Starting from Kotlin 1.4 dependency on the standard library added by default)
packagename is changed
handle error of server response

Helgapolis committed:
gradle version is updated, library versions are updated,
usecases in tracklist/domain are moved to tracklist/domain/usecase
compose library is added
ViewModelProvider.Factory is deprecated, replaced to ViewModelProvider.AndroidViewModelFactory

Helgapolis committed:
compose, room, gson versions are updated

Helgapolis committed:
file is removed
